mcwb/volume.py

```diff
@@ -0,0 +1,131 @@
+from mcipc.rcon.enumerations import Item
+from mcipc.rcon.errors import InvalidArgument
+from mcipc.rcon.je import Client
+
+from mcwb import Anchor, Anchor3, Direction, Vec3, mktunnel
+
+__all__ = ['Volume']
+
+
+class Volume:
+    """
+    Describes a 3d space in a Minecraft world using a starting point and
+    size. The starting point can be any vertex or the middle of one of the
+    horizontal faces, using the cardinal terminology defined in Anchor3.
+    """
+
+    def __init__(
+        self,
+        position: Vec3,
+        size: Vec3 = None,
+        anchor: Anchor3 = Anchor3.BOTTOM_NW,
+        end: Vec3 = None,  # opposite corner instead of size and anchor
+    ):
+        if (size is None) == (end is None):
+            raise InvalidArgument(
+                "Volume constructor takes only one of size or end")
+        position = Vec3(*position)
+
+        if end is None:
+            size = Vec3(*size)
+            offset = Vec3(0, 0, 0)
+            if anchor.value in Anchor3.TOP.value:
+                offset += Vec3(0, 1 - size.y, 0)
+            if anchor.value in Anchor3.SOUTH.value:
+                offset += Vec3(0, 0, 1 - size.z)
+            if anchor.value in Anchor3.EAST.value:
+                offset += Vec3(1 - size.x, 0, 0)
+            if anchor.value in Anchor3.MIDDLE_FACE.value:
+                offset -= Vec3(int(size.x / 2), 0, int(size.z / 2))
+            elif anchor.value == Anchor3.MIDDLE.value:
```

Comment:
The `else` here comes as a surprise in so far, as that it resets the `offset` from all possible `if`s above.
Maybe check this first?

Reply:
I think that this is correct, MIDDLE is the special case which does not appear in any of the faces (now defined in Anchor3Face) so none of the other if clauses would have executed for MIDDLE.

This is simply because MIDDLE does not align with any face of the volume (wheraa  MIDDLE_TOP and MIDDLE_BOTTOM do.

Reply:
however this leads me to think that the new Anchor3Face.MIDDLE should be renamed to .MIDDLE_FACE otherwise this is confusing.